projects/showcase/src/app/helper/map-testing-helper.directive.ts

```diff
@@ -0,0 +1,45 @@
+import { Directive, HostListener, OnDestroy, OnInit } from '@angular/core';
+import { MapComponent } from 'ngx-maplibre-gl';
+
+@Directive({
+  // tslint:disable-next-line: directive-selector
+  selector: 'mgl-map',
+})
+export class MapTestingHelperDirective implements OnInit, OnDestroy {
+  // Since the Angular environment is production by default, our e2e tests don't
+  // have a good way to reference the map component or underlying MapLibre-GL instance
+  // in order to check the state of things beyond the DOM. This directive adds some
+  // behavior to expose information as a global object that tests can access.
+
+  constructor(private map: MapComponent) {
+    if ((window as any).Cypress != null) {
+      console.info('Cypress detected. Setting preserveDrawingBuffer=true');
+      this.map.preserveDrawingBuffer = true;
+    }
+  }
+
+  @HostListener('idle', ['$event']) onMapIdle(data: any) {
+    (window as any).mglMapTestHelper.map = data.target;
```

Comment:
Is there a way to do it without "as any"?
I've seen that cypress now supports component testing...?

Reply:
Maybe it can be defined by means of typings, but I tried adding it that way (in a *.d.ts module) and I couldn't get it to work. May also be able to hold the `any` object in a variable and refer to it instead of repeatedly casting.

I still don't know a lot about Cypress and its capabilities as far as component testing, but when the angular environment is production (which it is now by default and as defined in angular.json), then ngDevMode is false and you don't get a global `ng` object or any other way to get at the code. That was the problem that this simple directive tries to overcome so that tests could have access to some otherwise internal data.

Reply:
Yea, I also tried to think about a solution, but the concept behind e2e tests is that you shouldn't be aware of the internals.
What I had in mind is some kind of visual effect to let the user (and the tests) know when map finished loading.
But I didn't had a chance to implement it.